File: #156_Binary_Tree_Upside_Down/Solution.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

public class Solution {
	TreeNode newRoot;
    public TreeNode upsideDownBinaryTree(TreeNode root) {
		if (root == null) return null;
		postOrder(root);
		root.left = null;
		root.right = null;
		return newRoot;
    }

    private TreeNode postOrder(TreeNode root) {
    	if (root.left == null) {
    		newRoot = root;
    		return newRoot;
    	}
    	TreeNode cur = postOrder(root.left);
    	cur.left = root.right;
    	cur.right = root;
    	return root;
    }
}

File: #156_Binary_Tree_Upside_Down/Solution.py
# Definition for a binary tree node.
class TreeNode:
	def __init__(self, x):
		self.val = x
		self.left = None
		self.right = None

class Solution:
	# @param {TreeNode} root
	# @return {TreeNode}
	def upsideDownBinaryTree(self, root):
		if not root or not root.left:
			return root
		newRoot = self.upsideDownBinaryTree(root.left)
		root.left.left = root.right
		root.left.right = root
		root.left = None
		root.right = None
		return newRoot